arch: riscv: introduce common layer for RISC-V custom CSRs

jimmyzhe opened the pull request:
This PR is inspired by #91940.

RISC-V allows custom CSR implementation. Some SoCs from different SoC vendors may share the same core or the same custom CSR definitions and drivers.
For example, Andes AE350, Telink TLSR9518, and Egis ET171 all support Andes-specific CSRs.

This PR introduces ~~`soc/common/riscv-custom`~~ `zephyr/arch/riscv/custom`, which centralizes support for custom CSRs, allowing code reuse across SoCs that come from different vendors.

Currently supported these custom CSR:
1. Andes
2. Nuclei
3. OpenHWGroup CVA6
4. OpenISA RI5CY
5. OpenISA Zero-riscy
6. T-Head

jimmyzhe committed:
arch: riscv: introduce RISC-V custom CSR common layer

RISC-V allows custom CSR implementation. Some SoCs from different vendors
may share the same core or the same custom CSR definitions and drivers.

This patch introduces 'arch/riscv/custom', which centralizes support
for custom CSRs, allowing code reuse across SoCs that come from different
vendors. Currently supported thess custom CSR:
  1. Andes
  2. Nuclei
  3. OpenHWGroup CVA6
  4. OpenISA RI5CY
  5. OpenISA Zero-riscy
  6. T-Head Xuantie

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
arch: riscv: custom: add Andes CSR support

Rework Andes-specific CSR to use RISC-V custom CSR common code.
Move these stuff to 'arch/riscv/custom/andes':

1. Rename 'soc_v5.h' to 'andes_csr.h' for CSR definitions.
2. Replace '_start' with '__reset' hook for low-level CSR initialization.
3. Move CSR context to common macro '__custom_csr_save/restore_context'.
4. Move 'EXECIT' CSR support to common code.
5. Move PMA CSR driver to common code.
6. Use RISC-V common linker.ld instead of SoC-specific linker.ld.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
soc: telink: tlsr951x: use RISC-V custom CSR common code

TLSR951x also supports Andes extended CSR. Reworks the following CSR
handling to use the RISC-V custom CSR common code:

1. Use common macros for HWDSP CSR context save/restore.
2. Use common macros for PFT CSR context save/restore.
3. Use common low-level CSR initialization via __reset hook.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
arch: riscv: : custom: add Nuclei CSR support

GD32VF103 uses Nuclei-specific CSR. Move 'nuclei_csr.h' to
'arch/riscv/custom' to allow reuse across SoCs with the same Nuclei core.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
arch: riscv: custom: add OpenHW Group CVA6 CSR support

CVA6 supports custom CSR. Move 'cva6.h' to 'arch/riscv/custom/cva6_csr.h',
allowing other SoCs using the CVA6 core to reuse the same CSR definitions.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

jimmyzhe committed:
arch: riscv: custom: add OpenISA RI5CY and Zero-RISCY CSR support

Update OpenISA RI5CY and Zero-RISCY CSR handling to use RISC-V custom
CSR common code. Move these stuff to 'arch/riscv/custom/openisa':

1. Rename 'soc_ri5cy.h' to 'ri5cy_csr.h' for CSR definitions.
2. Rename 'soc_zero_riscy.h' to 'zero_riscy_csr.h' for CSR definitions.
3. Move CSR context to common macro '__custom_csr_save/restore_context'.
4. Move compiler option '-march=rv32imcxpulpv2' to common code.

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

nordicjm commented on soc/common/riscv-custom/andes/Kconfig:
```diff
@@ -0,0 +1,66 @@
+config CPU_HAS_ANDES_EXECIT
+	bool
+	help
+		The AndesCore supports EXEC.IT instruction.
```
help text indent is 1x tab followed by 2x spaces

jimmyzhe replied:
Fixed it, thanks.

nordicjm commented on soc/common/riscv-custom/andes/Kconfig:
```diff
@@ -0,0 +1,66 @@
+		The AndesCore supports EXEC.IT instruction.
+
+config RISCV_CUSTOM_CSR_ANDES_EXECIT
+	bool "Andes V5 EXEC.IT extension"
```
why do any of these have prompts?

jimmyzhe commented:
Removed old Kconfig option in adp_xc7k_ae350_clic_defconfig.

Hi @kevinwang821020 , there are some issues with the [dma_andes_atcdmac300](https://github.com/zephyrproject-rtos/zephyr/blob/64131856199342d0cb6507e5c57355c82f24dd69/drivers/dma/dma_andes_atcdmac300.c#L403) driver check or the [adp_xc7k_ae350.conf](https://github.com/zephyrproject-rtos/zephyr/blob/64131856199342d0cb6507e5c57355c82f24dd69/tests/drivers/dma/loop_transfer/boards/adp_xc7k_ae350.conf#L1) in tests/drivers/dma.
Could you help take a look?

kevinwang821020 commented:
> Removed old Kconfig option in adp_xc7k_ae350_clic_defconfig.
> 
> Hi @kevinwang821020 , there are some issues with the [dma_andes_atcdmac300](https://github.com/zephyrproject-rtos/zephyr/blob/64131856199342d0cb6507e5c57355c82f24dd69/drivers/dma/dma_andes_atcdmac300.c#L403) driver check or the [adp_xc7k_ae350.conf](https://github.com/zephyrproject-rtos/zephyr/blob/64131856199342d0cb6507e5c57355c82f24dd69/tests/drivers/dma/loop_transfer/boards/adp_xc7k_ae350.conf#L1) in tests/drivers/dma. Could you help take a look?

Based on the current atcdmac driver codebase, it need to do the cache operations(CONFIG_CACHE_MANAGEMENT) for chain_block when DCACHE is enabled, even CONFIG_NOCACHE_MEMORY is enable.

I have already prepared a previous PR ( PR #85272 ) which includes fixing this problem.

I'd appreciate it if someone could help review that.

jimmyzhe commented:
> Either this is vendor specific in which case this should be in `zephyr/soc/<vendor>/` _or_ this is more architectural in which case it might rather be in `zephyr/arch/riscv/custom/`.

Right, custom CSR is RISC-V feature, I think placing these in `zephyr/arch/riscv/custom` is more appropriate and simple to reuse.

Moved the custom CSR common code to `zephyr/arch/riscv/custom` and also fixed the help text indentation in `zephyr/arch/riscv/custom/andes/Kconfig`.

jimmyzhe commented:
> 
> I have already prepared a previous PR ( PR #85272 ) which includes fixing this problem.

Rebased to pass CI.

fkokosinski reviewed (dismissed):
Hi, thanks for this! LGTM

nandojve commented on soc/andestech/ae350/Kconfig:
```diff
@@ -11,11 +11,19 @@ config SOC_SERIES_ANDES_AE350
 	select RISCV_ISA_EXT_ZIFENCEI
-	select ATOMIC_OPERATIONS_BUILTIN
-	select INCLUDE_RESET_VECTOR
+	select RISCV_CUSTOM_CSR
+	select RISCV_CUSTOM_CSR_ANDES
```
Why RISCV_CUSTOM_CSR_FOO do no select RISCV_CUSTOM_CSR ?

jimmyzhe commented on arch/riscv/custom/Kconfig:
```diff
@@ -0,0 +1,34 @@
+
+# Configuration options for RISC-V cores with custom CSR support
+
+config RISCV_CUSTOM_CSR_ANDES
```
I originally addedd `RISCV_CUSTOM_CSR_FOO` to handle AMP platform may enable 2 cores with different custom CSR, like rv32m1.
https://github.com/zephyrproject-rtos/zephyr/blob/0a339a2d7b6df61af7054e89cea24d215b0e9c8e/dts/riscv/openisa/rv32m1.dtsi#L25
https://github.com/zephyrproject-rtos/zephyr/blob/0a339a2d7b6df61af7054e89cea24d215b0e9c8e/dts/riscv/openisa/rv32m1.dtsi#L32

But after check the current RISC-V AMP platforms. each board interface only exists one `DT_HAS_VENDOR_CORE_ENABLED` in the final configuration, like
- rv32m1_vega_openisa_rv32m1_ri5cy -> `DT_HAS_OPENISA_RI5CY_ENABLED`
- rv32m1_vega_openisa_rv32m1_zero_riscy -> `DT_HAS_OPENISA_ZERO_RI5CY_ENABLED`

So these `RISCV_CUSTOM_CSR_FOO` can be replaced by `DT_HAS_VENDOR_CORE_ENABLED`.
I will remove these symbols.

jimmyzhe commented on arch/riscv/custom/andes/execit.ld:
```diff
@@ -1,6 +1,5 @@
 /*
- * Copyright (c) 2023 Andes Technology Corporation
- *
+ * Copyright (c) 2025 Andes Technology Corporation
```
Right, I only moved this file’s directory. I will restore this.

jimmyzhe commented:
Update:
1. Replaced  all `RISCV_CUSTOM_CSR_FOO` by `DT_HAS_VENDOR_CORE_ENABLED`
2. [arch/riscv/custom/andes/execit.ld](https://github.com/zephyrproject-rtos/zephyr/compare/f51a12450be84b23aece0e5dc26302a008606ac1..8037d9e4518b91dc0e17cda72e8e0eeb8e247fca#diff-c23093ec0a1fac8eb6d6dbedb9e91bf7d7db3778e150e3d91be3972751de0d06): removed redundant copyright change
3. [arch/riscv/custom/andes/csr_context.h](https://github.com/zephyrproject-rtos/zephyr/compare/f51a12450be84b23aece0e5dc26302a008606ac1..8037d9e4518b91dc0e17cda72e8e0eeb8e247fca#diff-2080b11398669dd09cd4f1de40a2b6ef0bf66dec18eadaa24025a5a812478e74): fixed code typo
4. [drivers/cache/Kconfig.andes](https://github.com/zephyrproject-rtos/zephyr/compare/f51a12450be84b23aece0e5dc26302a008606ac1..8037d9e4518b91dc0e17cda72e8e0eeb8e247fca#diff-17cdbc946576621bdfdf3e786a4656fdc5b355c77612026c09cda742a272e5b2): refined description

jimmyzhe commented:
Thanks. Rebased.

jimmyzhe commented on arch/riscv/custom/andes/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+
+# Note: AndeStar V5 DSP needs custom Andes V5 toolchain
+if(CONFIG_RISCV_CUSTOM_CSR_ANDES_HWDSP)
+  zephyr_cc_option(-mext-dsp)
```
Thanks for your earlier hint, I add support for Andes toolchain in #96833.

In this PR, I keep Andes-specific compiler, linker flags in the original file (soc/andestech/ae350/CMakeLists.txt), and will be moved in #96833.

jimmyzhe commented:
Rebased to main and add the following change:
1. Add `config RISCV_CUSTOM_CSR_ANDES_NEXECIT` for the new Andes EXECIT instruction encoding selection.
2. Keep the Andes-specific compiler, linker flags in original soc file. These flags will be centralize support in cmake/toolchain/andes in #96833.

jimmyzhe committed:
soc: egis: et171: add support for Andes custom CSRs

Egis ET171 implements Andes custom CSRs. Enable the following features:

1. Low level initialization of Andes CSRs
2. HWDSP and PowerBrake extensions with context save/restore
3. EXEC.IT extension

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>

ycsin reviewed (dismissed):
LGTM otherwise

jimmyzhe committed:
arch: riscv: custom: add T-Head Xuantie CSR support

Move Xuantie supprot from arch/riscv/core/xuantie to the custom common
layer arch/riscv/custom/thead, with the following changes:

1. Rename Kconfig name
   CACHE_XTHEADCMO -> RISCV_CUSTOM_CSR_THEAD_CMO
2. Split the original arch/riscv/core/xuantie/Kconfig to
   a. arch/riscv/custom/thead/Kconfig: for T-Head extension
   b. arch/riscv/custom/thead/Kconfig.core: for T-Head CPU series
      (e.g. Xuantie E907)
3. Move cache line size defaults to SoC devicetree

Signed-off-by: Jimmy Zheng <jimmyzhe@andestech.com>
